cclib/method/ddec.py

```diff
@@ -0,0 +1,170 @@
+from cclib.parser.utils import find_package
+
+
+class MissingInputError(Exception):
+    pass
+
+
+class DDEC6(Method):
+    """DDEC6 charges."""
+
+    # All of these are required for DDEC6 charges.
+    required_attrs = ("homos", "mocoeffs", "nbasis", "gbasis")
+
+    def __init__(
+        self, data, volume, proatom_path=None, progress=None, loglevel=logging.INFO, logname="Log"
+    ):
+        super(DDEC6, self).__init__(data, progress, loglevel, logname)
+
+        self.volume = volume
+        self.fragresults = None
+
+        if numpy.sum(self.data.coreelectrons) != 0:
+            # TODO: Pseudopotentials should be added back
+            pass
+
+        assert os.path.isdir(
+            proatom_path
+        ), "Directory that contains chargemol proatom densities should be added as an input."
+        # Read in reference charges
+        self.proatom_density = []
+        self.radial_grid_r = []
+
+        for atom_number in self.data.atomnos:
+            density, r = self._read_proatom(proatom_path, atom_number, 0)
+
+            self.proatom_density.append(density)
+            self.radial_grid_r.append(r)
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "DDEC6 charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "DDEC6({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(DDEC6, self)._check_required_attributes()
+
+    def _read_proatom(self, directory, atom_num, charge):
+        """Return a list containing proatom reference densities."""
+        # TODO: Treat calculations with psuedopotentials
+        # File name format:
+        #   ** Chargemol **
+        #       c2_[atom number]_[nuclear charge]_[electron count]_[cutoff radius]_[# shells]
+        #   ** Horton **
+        #       atoms.h5
+        # File format:
+        #   Starting from line 13, each line contains the charge densities for each shell
+        chargemol_path = os.path.join(
+            directory,
+            "c2_{:03d}_{:03d}_{:03d}_500_100.txt".format(atom_num, atom_num, atom_num - charge),
+        )
+        horton_path = os.path.join(directory, "atoms.h5")
+
+        if os.path.isfile(chargemol_path):
+            # Use chargemol proatom densities
+            # Each shell is 5.600570811644798 angstroms apart (uniform).
+            # *scalefactor* = 10.58354497764173 in module_global_parameter.f08
+            density = numpy.loadtxt(path, skiprows=12, dtype=float)
+            radiusgrid = numpy.arange(1, len(density + 1)) * 5.600570811644798
+
+        elif os.path.isfile(horton_path):
+            # Use horton proatom densities
+            _found_h5py = find_package("h5py")
+            assert _found_h5py, ("h5py is needed to read in proatom densities from horton.")
+
+            import h5py
+
+            with h5py.File(horton_path, "r") as proatomdb:
```

Comment:
It looks like there's one file in the Horton case. Do we need do read it multiple times?

Reply:
horton indeed has only one file and it would be more efficient to read in once in `__init__` and store it somewhere for future reference. Can I revisit this point after finishing up rest of the DDEC algorithms though? I am just not sure yet in which form proatomic densities will be needed in later steps. I'll add a `TODO` comment referring to this at the top of this code block. Thanks!

Reply:
SG, excllent plan.